subsys/usb/device_next/class/usbd_dfu.c

```diff
@@ -0,0 +1,794 @@
+		errno = -ENOTSUP;
+		dfu_error(c_data, DFU_ERROR, ERR_UNKNOWN);
+	}
+
+	return 0;
+}
+
+static int dfu_mode_control_to_host(struct usbd_class_data *const c_data,
+				    const struct usb_setup_packet *const setup,
+				    struct net_buf *const buf)
+{
+	struct usbd_dfu_data *data = usbd_class_get_private(c_data);
+
+	errno = dfu_set_next_state(c_data, setup);
+
+	if (errno == 0) {
+		switch (setup->bRequest) {
+		case USB_DFU_REQ_GETSTATUS:
+			errno = handle_get_status(c_data, setup, buf);
+			break;
+		case USB_DFU_REQ_GETSTATE:
+			errno = handle_get_state(c_data, setup, buf);
+			break;
+		case USB_DFU_REQ_UPLOAD:
+			errno = handle_upload(c_data, setup, buf);
+			break;
+		default:
+			break;
+		}
+	}
+
+	data->state = data->next;
+
+	return 0;
+}
+
+static int dfu_mode_control_to_dev(struct usbd_class_data *const c_data,
+				   const struct usb_setup_packet *const setup,
+				   const struct net_buf *const buf)
+{
+	struct usbd_dfu_data *data = usbd_class_get_private(c_data);
+
+	errno = dfu_set_next_state(c_data, setup);
+
+	if (errno == 0) {
+		if (setup->bRequest == USB_DFU_REQ_DNLOAD) {
+			handle_download(c_data, setup, buf);
+		}
+	}
+
+	data->state = data->next;
+
+	return 0;
+}
+
+static void dfu_mode_update(struct usbd_class_data *const c_data,
+			    const uint8_t iface, const uint8_t alternate)
+{
+	struct usbd_dfu_data *data = usbd_class_get_private(c_data);
+
+	LOG_DBG("Instance %p, interface %u alternate %u changed",
+		c_data, iface, alternate);
+
+	data->alternate = alternate;
+	data->image = NULL;
+
+	STRUCT_SECTION_FOREACH(usbd_dfu_image, image) {
+		if (image->if_desc->bAlternateSetting == alternate) {
+			data->image = image;
+			break;
+		}
+	}
+}
+
+static void *dfu_mode_get_desc(struct usbd_class_data *const c_data,
+			       const enum usbd_speed speed)
+{
+	struct usbd_dfu_data *data = usbd_class_get_private(c_data);
+
+	return data->dfu_mode_descs;
+}
+
+static int dfu_mode_init(struct usbd_class_data *const c_data)
+{
+	struct usbd_dfu_data *data = usbd_class_get_private(c_data);
+	struct usbd_context *uds_ctx = usbd_class_get_ctx(c_data);
+
+	LOG_DBG("Init class instance %p", c_data);
+	data->dfu_mode = true;
+	data->alternate = 0;
+	data->image = NULL;
+	data->state = DFU_IDLE;
+	data->next = DFU_IDLE;
+	data->ctx = usbd_class_get_ctx(c_data);
+
+	STRUCT_SECTION_FOREACH(usbd_dfu_image, image) {
+		if (image->if_desc->bAlternateSetting == data->alternate) {
+			data->image = image;
+		}
+
+		if (usbd_add_descriptor(uds_ctx, image->sd_nd)) {
+			LOG_ERR("Failed to add string descriptor");
+		} else {
+			image->if_desc->iInterface = usbd_str_desc_get_idx(image->sd_nd);
+		}
+	}
+
+	return data->image == NULL ? -EINVAL : 0;
+}
+
+struct usbd_class_api dfu_api = {
+	.control_to_host = dfu_mode_control_to_host,
+	.control_to_dev = dfu_mode_control_to_dev,
+	.update = dfu_mode_update,
+	.get_desc = dfu_mode_get_desc,
+	.init = dfu_mode_init,
+};
+
+USBD_DEFINE_CLASS(dfu_dfu, &dfu_api, &dfu_data, NULL);
```

Comment:
Non-blocking wish:

It would be nice to be able to only enable the run-time part, allowing an application to delegate the actual DFU part to a USB DFU-capable bootloader. This would especially be handy for smaller MCUs with a built-in ROM USB DFU bootloader.

As far as I can tell, this could be done with just a boolean Kconfig for enabling the actual DFU part (default y). What do you think?

Reply:
Sounds good. We can follow-up with a PR for allowing just the run-time part at a later point in time.